hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -972,11 +1001,44 @@ private void doBatchOp(final RegionActionResult.Builder builder, final HRegion r
 
       OperationStatus[] codes = region.batchMutate(mArray, atomic, HConstants.NO_NONCE,
         HConstants.NO_NONCE);
+      if (atomic) {
+        List<ResultOrException> resultOrExceptions = new ArrayList<>();
+        List<Result> results = new ArrayList<>();
+        for (i = 0; i < codes.length; i++) {
+          if (codes[i].getResult() != null) {
+            results.add(codes[i].getResult());
+          }
+          if (i != 0) {
+            resultOrExceptions.add(getResultOrException(
+              ClientProtos.Result.getDefaultInstance(), i));
+          }
+        }
+
+        if (results.isEmpty()) {
+          resultOrExceptions.add(0, getResultOrException(
+            ClientProtos.Result.getDefaultInstance(), 0));
+        } else {
+          // Set the result of the Increment/Append operations to the first element of the
+          // ResultOrException list
+          List<Cell> cellList = new ArrayList<>();
```

Comment:
Mind explaining a bit about the logic here?  I think batchMutate is for multiple operations? Why here we could merge the cells from different Results? We can make sure they are always for the same row?

Reply:
batchMutate can also be used by mutateRow(), and the case that atomic is true is when a client calls mutateRow(). In this case, of course, the results are always for the same row, and we need to return the results of Increment/Append operations as one Result object. That's why we need to merge the cell from different Results. And this logic sets the result to the first element of the ResultOrException list. Thanks.

Reply:
Seems my comment is disappeared, maybe I forgot to push the comment button. The concern here is how we can make sure that when arriving this branch, we are going to treat the result for mutateRow. Here we are in doBatchOp method, which will be called in RSRpcService.multi method, and the atomic flag is read from the protobuf message. This means the only place where we set atomic to true is for mutateRow? Maybe this is the case as for atomic batching we need to use the MultiRowMutationEndpoint, just asking. And I suggest that if this is the case then we should add more comments in the code to explicitly say this so other developers will not be confusing.

Reply:
> Here we are in doBatchOp method, which will be called in RSRpcService.multi method, and the atomic flag is read from the protobuf message. This means the only place where we set atomic to true is for mutateRow? Maybe this is the case as for atomic batching we need to use the MultiRowMutationEndpoint, just asking.

Yes, the only place where we set atomic to true is for mutateRow. In case of MultiRowMutationEndpoint, it calls Region.mutateRowsWithLocks() directly. So it doesn't call the doBatchOp method.

> I suggest that if this is the case then we should add more comments in the code to explicitly say this so other developers will not be confusing.

Sure I will add more comments. Thanks.